File: backend-spring-boot/src/test/java/no/vicx/backend/jwt/github/vm/GitHubUserVmTest.java
package no.vicx.backend.jwt.github.vm;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

class GitHubUserVmTest {

    @Test
    void isEmpty_givenEmptyUserVm_expectTrue() {
        assertTrue(new GitHubUserVm(null,null,null,null,null).isEmpty());
    }

    @ParameterizedTest
    @MethodSource("provideNonEmptyUsers")
    void isEmpty_givenNonEmptyUserVm_expectFalse(GitHubUserVm userVm) {
        assertFalse(userVm.isEmpty());
    }

    private static Stream<Arguments> provideNonEmptyUsers() {
        return Stream.of(
                Arguments.of(
                        new GitHubUserVm(
                                "~id~", "~login~", "~user1~", "~email~", "~avatar~")),
                Arguments.of(
                        new GitHubUserVm(
                                null, "~login~", "~user1~", "~email~", "~avatar~")),
                Arguments.of(
                        new GitHubUserVm(
                                null, null, "~user1~", "~email~", "~avatar~")),
                Arguments.of(
                        new GitHubUserVm(
                                null, null, null, "~email~", "~avatar~")),
                Arguments.of(
                        new GitHubUserVm(
                                null, null, null, null, "~avatar~"))
        );
    }
}

File: backend-spring-boot/src/test/kotlin/no/vicx/backend/jwt/github/vm/GitHubUserVmTest.kt
package no.vicx.backend.jwt.github.vm

import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.data.Row1
import io.kotest.data.forAll
import io.kotest.matchers.shouldBe

class GitHubUserVmTest :
    BehaviorSpec({

        Given("a GitHubUserVm with all values set to empty") {
            val sut = GitHubUserVm("", "", "", "", "")

            When("calling isEmpty") {
                val isEmpty = sut.isEmpty

                Then("It should be empty") {
                    isEmpty shouldBe true
                }
            }
        }

        Given("a GitHubUserVm with at least one value") {
            forAll(
                Row1(GitHubUserVm("~id~", "", "", "", "")),
                Row1(GitHubUserVm("", "~login~", "", "", "")),
                Row1(GitHubUserVm("", "", "~name~", "", "")),
                Row1(GitHubUserVm("", "", "", "~email~", "")),
                Row1(GitHubUserVm("", "", "", "", "~avatar~")),
            ) { sut ->

                When("calling isEmpty: $sut") {
                    val isEmpty = sut.isEmpty

                    Then("It should not be empty") {
                        isEmpty shouldBe false
                    }
                }
            }
        }
    })
